### PR TITLE
Sections with other features

### LINKED ISSUE #47621: Ensure other features interact well with Sections
Now that we're making `SectionContent`/`SectionOutlet` into public features with .NET 8, we need to ensure they interact well with other features. Whatever behaviors exist when .NET 8 ships will become something we're fairly committed to.

Based on how sections are implemented (i.e., `SectionContent` passes a `RenderFragment` to a `SectionOutlet` in some other part of the component hierarchy), I would expect this to cause some unintuitive behaviors in the following features:

 * **Cascading parameters**: the section content looks as if it would inherit cascading parameters placed around the `<SectionContent>`, but it will actually get ones that are placed around the `<SectionOutlet>` (since the renderfragment is executed by the `SectionOutlet`)
 * **Error boundaries**: if there's an unhandled exception, you'd expect it to be caught by an `<ErrorBoundary>` placed around the `<SectionContent>`, but it will actually get caught by one placed around the `<SectionOutlet>` (same reason)
 * **Streaming rendering**: you'd expect the "streaming-or-not" flag to be determined by the component supplying the content for a `<SectionContent>`, but it will actually be determined by the component rendering the `<SectionOutlet>`

### Underlying issue

The core issue is that each of those features walk up the chain of parent components in the hierarchy to determine how to behave. Since the `RenderFragment` is executed by the `SectionOutlet`, the components rendered there are descendants of the `SectionOutlet` not the `SectionContent` - even though in the app source code, it looks as if they are descendants of the `SectionContent`.

### Proposed solution

This is just a suggestion. I haven't actually implemented this to fully validate it.

 * Update `ComponentState` so that as well as tracking the `ParentComponentState`, it also has a `LogicalParentComponentState` property.
   * By default this is always just set to the same value as `ParentComponentState`
   * ... but for `SectionOutlet` in particular, it has a special behavior of being set to the component state for the current `SectionContent`. That is, whenever `SectionOutlet` switches to a new source for its renderfragment, it also changes its own `LogicalParentComponentState` to match.
     * This is a magic built-in behavior of `SectionOutlet` provided by the framework, since sections are a special framework feature. No other component needs to be able to do this.
 * Change each of these features so that, when walking the list of parent components to find ancestors, they walk up the chain of `LogicalParentComponentState` values instead of `ParentComponentState`:
   * Finding cascading value suppliers
   * Finding error boundary suppliers
   * Deciding whether or not to run in streaming rendering mode
* Add E2E tests showing each of those features now behave in an intuitive way when used with sections
* Remove [this line](https://github.com/dotnet/aspnetcore/blob/main/src/Components/Components/src/Sections/SectionOutlet.cs#L9) that previously worked around this. This will make streaming rendering within `SectionContent` vary according to the host component, instead of being hardcoded to on in all cases.

When solving this issue, we should try to do so in a way that has zero perf impact for cases that aren't using sections. That's an aspect of the proposed solution above. So, for example, we should *not* have logic that runs on every component instantiation to check whether it's inside some `SectionContent` or similar.

### PR DESCRIPTION
**Pull request by surayya-MS**
# Sections with other features

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description

See [API Proposal](https://github.com/dotnet/aspnetcore/issues/48557) for complete description.

Fixes #47621

### CODE REVIEW

File: src/Components/Components/src/Rendering/ComponentState.cs [hunk 1]
```diff
@@ -61,8 +62,21 @@ public ComponentState(Renderer renderer, int componentId, IComponent component,
...
+    /// Gets the <see cref="ComponentState"/> of the logical parent component, or null if this is a root component.
+    /// </summary>
+    public ComponentState? LogicalParentComponentState { get; private set; }
+
     internal RenderTreeBuilder CurrentRenderTree { get; set; }
 
+    /// <summary>
+    /// Changes the <see cref="ComponentState"/> of the logical parent component.
+    /// </summary>
+    protected internal virtual void LogicalParentComponentStateChanged(ComponentState? logicalParentComponent)
```
Comment (javiercn) [inline src/Components/Components/src/Rendering/ComponentState.cs:75]: Can this be `private protected`?

File: src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/StreamingRendering.razor [hunk 2]
```diff
@@ -1,74 +1,7 @@
...
-        });
-
-        endpoints.MapGet(EndResponseUrl, () =>
-        {
-            StreamingDataChannel.Writer.Complete();
-            return "Response ended";
-        });
-    }
-}
+<NotEnabledStreamingRenderingComponent />
```
[the remark this answers is not in the document]
Comment (surayya-MS) [reply src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/StreamingRendering.razor:7]: I wanted to reuse `StreamingRendering.razor` component for the test [SectionsWithStreamingRendering.razor](https://github.com/dotnet/aspnetcore/pull/48518/files#diff-60f4ed82223f93173cf0e5fde0360af1922e87673502d978edd55abf4cdee24c) but `@attribute [StreamRendering(true)]` in  [StreamingRendering.razor](https://github.com/dotnet/aspnetcore/pull/48518/files#diff-c4fcbacd91fcd0d5787715ee729c70224001155ef92300922c193700affe9dfa) was a problem. 

The test suppose to check that "streaming-or-not" flag would be determined by the component supplying the content for `SectionContent`. Hence `StreamingRendering` attribute for the test should be in the parent of the `SectionContent`. 

[Here in StreamingRendering.razor](https://github.com/dotnet/aspnetcore/pull/48518/files#diff-c4fcbacd91fcd0d5787715ee729c70224001155ef92300922c193700affe9dfa) I moved everything into new component `<NotEnabledStreamingRenderingComponent />` and left the `@attribute [StreamRendering(true)]`. Then I reused `<NotEnabledStreamingRenderingComponent />` inside `SectionContent` and wrapped `SectionContent` in a component that just enables streaming rendering [StreamingRendering.razor](https://github.com/dotnet/aspnetcore/pull/48518/files#diff-c4fcbacd91fcd0d5787715ee729c70224001155ef92300922c193700affe9dfa).

File: src/Components/test/testassets/BasicTestApp/SectionsTest/SectionsWithCascadingParameters.razor [hunk 3]
```diff
@@ -0,0 +1,29 @@
+﻿@using Microsoft.AspNetCore.Components.Sections
+
+<SectionOutlet SectionName="page" />
```
[the remark this answers is not in the document]
Comment (surayya-MS) [reply src/Components/test/testassets/BasicTestApp/SectionsTest/SectionsWithCascadingParameters.razor:3]: I'll add these types of tests for cascading parameters and error boundary. But it is hard (or impossible) to do that for streaming rendering as the page is not interactive. If let's say I place a button that when clicked changes the SectionName, the click wouldn't do anything. What do you think? Am I missing a test scenario that could test that?

**Comment by SteveSandersonMS** [comment]
@surayya-MS Thanks for the reminder about this. I'm looking through it now.

File: src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/StreamingRendering.razor (hunk 2 above)
[the remark this answers is not in the document]
Comment (SteveSandersonMS) [reply src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/StreamingRendering.razor:7]: Another name could be `UsesStreamingRenderingButDoesNotEnableIt` - a very lengthy name but quite descriptive :)

TBH I don't mind either way.

**Review by SteveSandersonMS — APPROVED** [review]
Excellent! This is a tricky area so I'm really glad you were able to figure out all the details. And the simplified implementation of sections (eliminating various internal interfaces) is much cleaner too.

**Pull request merged**